src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -53,33 +55,48 @@ public override void Initialize (AnalysisContext context)
 					}
 				}
 			});
-			// TODO: fix reporting for generic type substitutions. This should happen not only for method invocations,
-			// but for any reference to an instantiated method or type.
-			context.RegisterOperationAction (context => {
-				var invocationOperation = (IInvocationOperation) context.Operation;
-				ProcessInvocationOperation (context, invocationOperation);
-			}, OperationKind.Invocation);
+			context.RegisterSyntaxNodeAction (context => {
+				ProcessGenericParameters (context);
+			}, SyntaxKind.GenericName);
 		}
 
-		static void ProcessInvocationOperation (OperationAnalysisContext context, IInvocationOperation invocationOperation)
+		static void ProcessGenericParameters (SyntaxNodeAnalysisContext context)
 		{
-			if (context.ContainingSymbol.IsInRequiresUnreferencedCodeAttributeScope ())
+			// We dont need to check for RUC for INamedTypeSymbol here
+			if (context.ContainingSymbol is not null
+				&& context.ContainingSymbol is not INamedTypeSymbol
+				&& context.ContainingSymbol.IsInRequiresUnreferencedCodeAttributeScope ())
 				return;
 
-			ProcessTypeArguments (context, invocationOperation);
-		}
+			ImmutableArray<ITypeParameterSymbol> typeParams = default;
+			ImmutableArray<ITypeSymbol> typeArgs = default;
+			if (context.SemanticModel.GetTypeInfo (context.Node).Type is INamedTypeSymbol type) {
+				// INamedTypeSymbol inside nameof, commonly used in [ExpectedWarning] can generate diagnostics
+				// Walking the node heirarchy to check if INamedTypeSymbol is inside a nameof
+				var parentNode = context.Node;
+				while (parentNode != null) {
+					if (parentNode is InvocationExpressionSyntax invocationExpression && invocationExpression.Expression is IdentifierNameSyntax ident1) {
+						if (ident1.Identifier.ValueText.Equals ("nameof"))
+							return;
+					}
```

Comment:
I think we should do something like
```C#
if (parentNode is InvocationExpressionSyntax invocationExpression && invocationExpression.Expression is IdentifierNameSyntax ident1) {
    if (ident1.Identifier.ValueText.Equals ("nameof"))
        return;
    break;
}					
```
In order to stop on the first InvocationExpression, it either was 'nameof' or it wasnt and we need to continue. Current code has the risk of continue computing and find an unrelated 'nameof' that disables the analysis

Reply:
Personally I would like to produce warnings even for generic types inside `nameof` - even though this might be noisier than the linker. I think it makes sense to think of the generic annotations as an extension of the type system, and to warn for any offending reference an instantiated type. It's a little painful for the linker repo but I don't know if it will be common in the wild.  I think this is similar to local variable declarations where (IIRC) the linker would not warn but the analyzer would. @vitek-karas do you have an opinion on this?

In any case, I think either approach is fine for now since it's easy enough to change. For a non-attribute testcase I think you could do something like this:

```csharp
class Outer<TOuter> {
  static void Method() {
    string typeName = nameof (TypeWithAnnotatedGenericParameter<TOuter>);
  }
}
```